Implement audience settings setup CTA

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8178 

## Relevant technical choices

This PR implements the audience settings setup CTA including its happy path behaviour.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

hussain-t commented on assets/js/modules/analytics-4/components/audience-segmentation/settings/SettingsCardVisitorGroups/SetupSuccess.js:
```diff
@@ -0,0 +1,93 @@
+		dismissNotificationForUser().then( () => navigateTo( dashboardURL ) );
+	}
+
+	if ( isDismissed ) {
```
Don't we need to show this setup success notification only if it was set up from the settings page?

nfmohit replied:
Interested to hear @techanvil's thoughts on this. We can simply dismiss the settings success notification when groups were enabled from the dashboard, so that this doesn't appear in the settings page.

techanvil replied:
Thanks @hussain-t and @nfmohit.

As Hussain has suggested, we should indeed only show this `SettingsCardVisitorGroups/SetupSuccess` notification when the the feature has been set up on the Settings page.

If the feature has been set up on the dashboard we'll show the `AudienceSegmentationSetupSuccessSubtleNotification` notification and there's no need or intention to show the `SettingsCardVisitorGroups/SetupSuccess` one when we subsequently land on the Settings page.

Similarly, we should not show the `AudienceSegmentationSetupSuccessSubtleNotification` notification when we next land on the dashboard after setting the feature up on the Settings page.

hussain-t replied:
Thanks for confirming, @techanvil. Should we fix it as part of this ticket or a separate one since it's already over the estimate?

techanvil replied:
Hey @hussain-t, as this is within the scope of the issue and indeed the intent of it, and the issue is still open, we should address it here even if it's over the estimate. Hopefully it's a quick fix and won't impact it too much.

hussain-t replied:
SGTM!

nfmohit replied:
Thank you folks. I have updated the PR accordingly.

hussain-t replied:
Thanks, @nfmohit.

hussain-t reviewed (changes requested):
Excellent work, @nfmohit. However, please consider the following:

1. I've left a few comments; please review them.
2. It seems the `SetupSuccess` notification should only be displayed if the audiences/visitors were enabled from the settings page. Currently, it shows even when set up from the dashboard. Similarly, if the setup is done from the dashboard, the `AudienceSegmentationSetupSuccessSubtleNotification` should display the settings setup notification. I would like to get @techanvil's input on this. If we decide to fix this, we could also address it in a separate issue. However, it’s better to clarify here.
3. Please resolve the merge conflicts.
4. While you have mentioned testing the dashboard setup CTA in the QAB, it would be better to specifically mention testing the unhappy paths introduced in [#8134](https://github.com/google/site-kit-wp/issues/8134).

nfmohit commented on assets/js/modules/analytics-4/hooks/useEnableAudienceGroup.js:
```diff
@@ -0,0 +1,151 @@
+		onError,
+	] );
+
+	// If the user ends up back on this component with the required scope granted,
```
Updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.js:
```diff
@@ -108,71 +103,14 @@ function AudienceSegmentationSetupCTAWidget( { Widget, WidgetNull } ) {
-	] );
+	const [ showErrorModal, setShowErrorModal ] = useState( false );
 
 	// If the user ends up back on this component with the required scope granted,
```
Whoops, this was an oversight, updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.js:
```diff
@@ -108,71 +103,14 @@ function AudienceSegmentationSetupCTAWidget( { Widget, WidgetNull } ) {
+	const { apiErrors, failedAudiences, isSaving, onEnableGroups } =
+		useEnableAudienceGroup( {
+			onError: () => {
+				setShowErrorModal( true );
```
Whoops, this was an oversight, updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.js:
```diff
@@ -108,71 +103,14 @@ function AudienceSegmentationSetupCTAWidget( { Widget, WidgetNull } ) {
-		setPermissionScopeError,
-		redirectURL,
-	] );
+	const [ showErrorModal, setShowErrorModal ] = useState( false );
```
Updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/settings/SettingsCardVisitorGroups/SetupSuccess.js:
```diff
@@ -0,0 +1,93 @@
+
+	if ( isDismissed ) {
+		return null;
+	}
```
Updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/settings/SettingsCardVisitorGroups/SetupCTA.js:
```diff
@@ -0,0 +1,70 @@
+
+	if ( isDismissed ) {
+		return null;
+	}
```
Updated, thanks!

nfmohit commented:
Thank you for your kind review on this, @hussain-t ! I've addressed all the points that you have raised, except for the one about showing or not showing the success notification in settings if groups were enabled in the main dashboard, where we are waiting for feedback from @techanvil.

techanvil commented:
Thanks @nfmohit and @hussain-t, I have replied to the question about the notifications, please see this https://github.com/google/site-kit-wp/pull/8954#discussion_r1680782051.

hussain-t reviewed (approved):
Excellent work, @nfmohit. LGTM 🎖️

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/settings/SettingsCardVisitorGroups/SetupSuccess.js:
```diff
@@ -0,0 +1,93 @@
+		);
+	}
+
+	function scrollToWidgetArea() {
+		dismissNotificationForUser().then( () => navigateTo( dashboardURL ) );
+	}
```
A minor point, but AFAICT there's no reason to use `.then()` here, this can use `async` / `await`, which would be more consistent with the wider codebase.

```suggestion
	async function scrollToWidgetArea() {
		await dismissNotificationForUser();
		navigateTo( dashboardURL );
	}
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/settings/SettingsCardVisitorGroups/SetupSuccess.js:
```diff
@@ -0,0 +1,94 @@
+					) }
+				</p>
+			</div>
+			<Button tertiary onClick={ dismissNotificationForUser }>
+				{ __( 'Got it', 'google-site-kit' ) }
+			</Button>
+			<Button onClick={ scrollToWidgetArea }>
+				{ __( 'Show me', 'google-site-kit' ) }
+			</Button>
```
These buttons should be wrapped in a `div.googlesitekit-subtle-notification__action` to ensure they wrap consistently with other "subtle" notifications:

```suggestion
			<div className="googlesitekit-subtle-notification__action">
				<Button tertiary onClick={ dismissNotificationForUser }>
					{ __( 'Got it', 'google-site-kit' ) }
				</Button>
				<Button onClick={ scrollToWidgetArea }>
					{ __( 'Show me', 'google-site-kit' ) }
				</Button>
			</div>
```

techanvil reviewed (approved):
LGTM, nice one @nfmohit!
